File: backend/app/models/domain/template.py
from typing import Any

from pydantic import BaseModel, Field, model_validator


class TemplateEntryFieldDef(BaseModel):
    key: str
    label: str
    field_type: str = "text"
    required: bool = False


class TemplateFieldDef(BaseModel):
    key: str
    label: str
    field_type: str = "text"
    required: bool = False
    section: str
    placeholder: str | None = None
    entry_fields: list[TemplateEntryFieldDef] = Field(default_factory=list)


class TemplateLayoutSection(BaseModel):
    section_key: str
    section_title: str
    layout: str
    order: int


class TemplateSchema(BaseModel):
    template_id: str
    name: str
    sections: list[str]
    fields: list[TemplateFieldDef]
    layout_sections: list[TemplateLayoutSection] = Field(default_factory=list)

    @model_validator(mode="after")
    def sort_layout_sections(self) -> "TemplateSchema":
        self.layout_sections = sorted(self.layout_sections, key=lambda item: item.order)
        return self


class TemplateFormData(BaseModel):
    template_id: str
    values: dict[str, Any] = Field(default_factory=dict)


class TemplateExtractionPayload(BaseModel):
    name: str
    sections: list[str]
    fields: list[TemplateFieldDef]
    layout_sections: list[TemplateLayoutSection]


class TemplateMappingPayload(BaseModel):
    values: dict[str, Any]


class FormatResponse(BaseModel):
    template_schema: TemplateSchema
    form_data: TemplateFormData
    resume_json: dict


class TemplatePdfRequest(BaseModel):
    form_data: TemplateFormData
    template_schema: TemplateSchema

File: frontend/types/template.ts
export interface TemplateEntryFieldDef {
  key: string;
  label: string;
  field_type: string;
  required: boolean;
}

export interface TemplateFieldDef {
  key: string;
  label: string;
  field_type: string;
  required: boolean;
  section: string;
  placeholder?: string | null;
  entry_fields: TemplateEntryFieldDef[];
}

export interface TemplateLayoutSection {
  section_key: string;
  section_title: string;
  layout: string;
  order: number;
}

export type TemplateFieldValue = string | Record<string, string | string[]>[];

export interface TemplateFormData {
  template_id: string;
  values: Record<string, TemplateFieldValue>;
}

export interface TemplateSchema {
  template_id: string;
  name: string;
  sections: string[];
  fields: TemplateFieldDef[];
  layout_sections: TemplateLayoutSection[];
}

export interface FormatResponse {
  template_schema: TemplateSchema;
  form_data: TemplateFormData;
  resume_json: Record<string, unknown>;
}

export interface TemplatePdfRequest {
  form_data: TemplateFormData;
  template_schema: TemplateSchema;
}

export interface FormatterState {
  templateSchema: TemplateSchema | null;
  formData: TemplateFormData | null;
}

export type TemplateEntryRecord = Record<string, string | string[]>;
